source/common/ssl/context_impl.cc

```diff
@@ -370,6 +381,142 @@ ServerContextImpl::ServerContextImpl(ContextManagerImpl& parent, Stats::Scope& s
                                },
                                this);
   }
+
+  if (!config.sessionTicketKeys().empty()) {
+    session_ticket_keys_.resize(config.sessionTicketKeys().size());
+    for (unsigned i = 0; i < config.sessionTicketKeys().size(); ++i) {
+      // If this changes, need to figure out how to deal with key files
+      // that previously worked.  For now, just assert so we'll notice that
+      // it changed if it does.
+      static_assert(sizeof(SessionTicketKey) == 80, "Input is expected to be this size");
+
+      const std::vector<uint8_t>& src_key = config.sessionTicketKeys().at(i);
+      SessionTicketKey& dst_key = session_ticket_keys_.at(i);
+
+      if (src_key.size() != sizeof(SessionTicketKey)) {
+        throw EnvoyException(fmt::format("Incorrect TLS session ticket key length.  "
+                                         "Index {}, length {}, expected length {}.",
+                                         i, src_key.size(), sizeof(SessionTicketKey)));
+      }
+
+      std::copy_n(src_key.begin(), dst_key.name.size(), dst_key.name.begin());
+      size_t pos = dst_key.name.size();
+      std::copy_n(src_key.begin() + pos, dst_key.hmac_key.size(), dst_key.hmac_key.begin());
+      pos += dst_key.hmac_key.size();
+      std::copy_n(src_key.begin() + pos, dst_key.aes_key.size(), dst_key.aes_key.begin());
+      pos += dst_key.aes_key.size();
+      ASSERT(src_key.begin() + pos == src_key.end());
+    }
+    SSL_CTX_set_tlsext_ticket_key_cb(
+        ctx_.get(),
+        [](SSL* ssl, uint8_t* key_name, uint8_t* iv, EVP_CIPHER_CTX* ctx, HMAC_CTX* hmac_ctx,
+           int encrypt) -> int {
+          ContextImpl* context_impl = static_cast<ContextImpl*>(
+              SSL_CTX_get_ex_data(SSL_get_SSL_CTX(ssl), sslContextIndex()));
+          return dynamic_cast<ServerContextImpl*>(context_impl)
+              ->sessionTicketProcess(ssl, key_name, iv, ctx, hmac_ctx, encrypt);
+        });
+  }
+
+  // Hash all the settings that affect whether the server will allow/accept
+  // the client connection.  This ensures that the client is always validated against
+  // the correct settings, even if session resumption across different listeners
+  // is enabled.
+  uint8_t session_context_buf[EVP_MAX_MD_SIZE] = {};
+  unsigned session_context_len = 0;
+  EVP_MD_CTX md;
+  int rc = EVP_DigestInit(&md, EVP_sha256());
+  RELEASE_ASSERT(rc == 1);
+
```

Comment:
Actually, thinking about this a bit more... I think we should also include server's certificate in the digest to prevent resumption across different certificates, even without any requirements on the client certificates.

Thoughts?

Reply:
One use case I'd like to support it resuming across certificates.  The use case for this: 

Front-proxy boxes come up, generate a certificate and CSR, get it signed, and now have a certificate unique to that box, and the private key never leaves the box.

Another use-case is when certificates are renewed.  If we can resume cross-certificate, there won't be a spike in handshakes caused by resumption failures.

Maybe we should hash in the CN and SANs?

Reply:
The solution for the renewals is actually pretty easy - hash SPKI instead of the certificate (this is used in certificate pinning, i.e. RFC7469, for example).

The unique certificate per box is a bit more tricky, and it seems that indeed, we need to either hash CN and SANs (perhaps also the Issuer?) or add the ability to configure Session Context ID.

Let's just hash CN and SANs (and Issuer?), and we can revisit SPKI / Session Context ID in the future, if there is any need for that.

Reply:
I've never dealt with the boringssl/openssl API for dealing with names.  I'm trying to use X509_get_subject_name() and X509_NAME_digest(), but cross-cert resumption fails.  I'm working on debugging the problem (trying to figure out what is included in X509_get_subject_name(); I suspect it includes the serial number, subject key identifier, or similar).  But if you happen to know of more appropriate calls to use, I'd appreciate any hints.

Either way, hopefully I'll have something put together on this soon.

Reply:
I think you could use:

    X509_NAME_print_ex(..., X509_get_subject_name(cert.get()), 0, XN_FLAG_RFC2253);

to be consistent with what we use in [`ConnectionImpl::subjectPeerCertificate()`](https://github.com/envoyproxy/envoy/blob/fa32989feee7e2eada48d5fc672c85eb8f42a834/source/common/ssl/connection_impl.cc#L262) and hash on that.

The result is `Subject=emailAddress=frontend-team@lyft.com,CN=Test Frontend Team,OU=Lyft Engineering,O=Lyft,L=San Francisco,ST=California,C=US`, so it shouldn't change across your certificates (unless you add some randomness to the subject to identify nodes).

Reply:
I was already well on the path to digging out just the CN.  What I ended up with never deals with the CN as a c-string, so it's nicer in that respect.  And it's similar to how the SANs are handled.  Let me know if you want me to switch to X509_NAME_print_ex(...) instead.